PeerDAS (#1)
hwwhww/consensus-specs: specs/_features/peerdas/p2p-interface.md

<!-- END doctoc generated TOC please keep comment here to allow auto update -->
<!-- /TOC -->

## Modifications in PeerDAS

### Preset

| Name                                     | Value                             | Description                                                         |
|------------------------------------------|-----------------------------------|---------------------------------------------------------------------|
| `KZG_COMMITMENTS_MERKLE_PROOF_INDEX`   | `uint64(get_generalized_index(BeaconBlockBody, 'blob_kzg_commitments'))` (= 27) | <!-- predefined --> Merkle proof index for `blob_kzg_commitments` |

### Containers

#### `DataColumnSidecar`

```python
class DataColumnSidecar(Container):
    index: LineIndex  # Index of column in extended matrix
    column: DataColumn
    kzg_commitments: List[KZGCommitment, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    kzg_proofs: List[KZGProof, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    signed_block_header: SignedBeaconBlockHeader
    kzg_commitment_merkle_proof: Vector[Bytes32, KZG_COMMITMENT_INCLUSION_PROOF_DEPTH]
```

#### `DataColumnIdentifier`

```python
class DataColumnIdentifier(Container):
    block_root: Root
    index: LineIndex

> ```

Author: This is the symmetric helper as `BlobIdentifier`. They are essentially the same data structure. 

If we can go modify Deneb spec, I'd like to add a general container in Deneb to replace both of them:
```python
class DataLineIdentifier(Container):
    block_root: Root
    index: LineIndex
```

@jacobkaufmann this is where `LineIndex` can be useful.

Reviewer: I think I prefer to have two different types for the identifier instead of just one, i.e. `DataColumnIdentifier` and `BlobIdentifier` instead of `DataLineIdentifier`, because blobs and columns are intrinsically different and it's better to be able to tell which one is being used whenever possible.

### Helpers

##### `verify_column_sidecar`

```python
def verify_column_sidecar(sidecar: DataColumnSidecar) -> bool:
    column = sidecar.column
    column_width = BYTES_PER_BLOB * 2 // NUMBER_OF_COLUMNS
    cell_count = len(column) // column_width
    cells = [column[i * column_width:(i + 1) * column_width] for i in range(cell_count)]

    assert len(cells) == len(sidecar.kzg_commitments) == len(sidecar.kzg_proofs)
    # KZG batch verify the cells match the corresponding commitments and proofs
    assert verify_cells(cells, sidecar.index, sidecar.kzg_commitments, sidecar.kzg_proofs)
    # Verify if it's included in the beacon block
    return is_valid_merkle_branch(
        leaf=hash_tree_root(data_line_sidecar.kzg_commitments),

>         branch=data_line_sidecar.kzg_commitments_merkle_proof,

Reviewer 2: ```suggestion
        branch=sidecar.kzg_commitments_merkle_proof,
```

        depth=floorlog2(KZG_COMMITMENTS_MERKLE_PROOF_INDEX),
        index=KZG_COMMITMENTS_MERKLE_PROOF_INDEX,

>         root=data_line_sidecar.signed_block_header.message.body_root,

Reviewer 2: ```suggestion
        root=sidecar.signed_block_header.message.body_root,
```

<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Modifications in PeerDAS](#modifications-in-peerdas)
  - [Preset](#preset)
  - [Containers](#containers)
    - [`DataColumnSidecar`](#datacolumnsidecar)
    - [`DataColumnIdentifier`](#datacolumnidentifier)
  - [Helpers](#helpers)
      - [`verify_column_sidecar`](#verify_column_sidecar)
  - [The gossip domain: gossipsub](#the-gossip-domain-gossipsub)
    - [Topics and messages](#topics-and-messages)
      - [Samples subnets](#samples-subnets)
        - [`data_column_sidecar_{subnet_id}`](#data_column_sidecar_subnet_id)
  - [The Req/Resp domain](#the-reqresp-domain)
    - [Messages](#messages)
      - [DataColumnSidecarByRoot v1](#datacolumnsidecarbyroot-v1)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->
<!-- /TOC -->

## Modifications in PeerDAS

### Preset

| Name                                     | Value                             | Description                                                         |
|------------------------------------------|-----------------------------------|---------------------------------------------------------------------|

> | `KZG_COMMITMENTS_MERKLE_PROOF_INDEX`   | `uint64(get_generalized_index(BeaconBlockBody, 'blob_kzg_commitments'))` (= 27) | <!-- predefined --> Merkle proof index for `blob_kzg_commitments` |

Reviewer 3: Match the name here to what is in line 48. one is called MERKLE_PROOF and one is called INCLUSION_PROOF

Author: Modified in 6f08051469d0237ef02139cc825c2ac2b6409855

    block_root: Root
    index: LineIndex
```

### Helpers

##### `verify_sample_proof_batch`

```python
def verify_sample_proof_batch(
        row_commitments: Sequence[KZGCommitment],
        row_ids: Sequence[LineIndex],
        column_ids: Sequence[LineIndex],
        datas: Sequence[Vector[BLSFieldElement, FIELD_ELEMENTS_PER_CELL]],
        proofs: Sequence[KZGProof]) -> bool:
    """
    Defined in polynomial-commitments-sampling.md
    """
    ...
```

##### `verify_column_sidecar`

```python
def verify_column_sidecar(sidecar: DataColumnSidecar) -> bool:
    column = sidecar.column
    cell_count = len(column) // FIELD_ELEMENTS_PER_CELL
    cells = [column[i * FIELD_ELEMENTS_PER_CELL:(i + 1) * FIELD_ELEMENTS_PER_CELL] for i in range(cell_count)]

    assert len(cells) == len(sidecar.kzg_commitments) == len(sidecar.kzg_proofs)
    # KZG batch verify the cells match the corresponding commitments and proofs
    assert verify_sample_proof_batch(
        row_commitments=sidecar.kzg_commitments,
        row_ids=list(range(sidecar.column)),  # all rows
        column_ids=[sidecar.index],
        datas=cells,
        proofs=sidecar.kzg_proofs
    )

>     # Verify if it's included in the beacon block

Reviewer 3: ```suggestion
    # Verify that the KZG commitments are correct with respect to the beacon block
```

- _[IGNORE]_ The sidecar is from a slot greater than the latest finalized slot -- i.e. validate that `block_header.slot > compute_start_slot_at_epoch(state.finalized_checkpoint.epoch)`
- _[REJECT]_ The proposer signature of `blob_sidecar.signed_block_header`, is valid with respect to the `block_header.proposer_index` pubkey.
- _[IGNORE]_ The sidecar's block's parent (defined by `block_header.parent_root`) has been seen (via both gossip and non-gossip sources) (a client MAY queue sidecars for processing once the parent block is retrieved).
- _[REJECT]_ The sidecar's block's parent (defined by `block_header.parent_root`) passes validation.
- _[REJECT]_ The sidecar is from a higher slot than the sidecar's block's parent (defined by `block_header.parent_root`).
- _[REJECT]_ The current finalized_checkpoint is an ancestor of the sidecar's block -- i.e. `get_checkpoint_block(store, block_header.parent_root, store.finalized_checkpoint.epoch) == store.finalized_checkpoint.root`.

> - _[REJECT]_ The sidecar's `kzg_commitments` field inclusion proof is valid as verified by `verify_data_column_sidecar_inclusion_proof(sidecar)`.

Author: These lines were primarily copied from BlobSidecar subnet validation. It's better to find a way to refactor and avoid duplication.

  If the `proposer_index` cannot immediately be verified against the expected shuffling, the sidecar MAY be queued for later processing while proposers for the block's branch are calculated -- in such a case _do not_ `REJECT`, instead `IGNORE` this message.

*Note:* In the `verify_data_column_sidecar_inclusion_proof(sidecar)` check, it should verify against the same set of `kzg_commitments` of the given beacon beacon. Client can choose to cache the result of the arguments tuple `(sidecar.kzg_commitments, sidecar.kzg_commitments_merkle_proof, sidecar.signed_block_header)`.

### The Req/Resp domain

#### Messages

##### DataColumnSidecarByRoot v1

**Protocol ID:** `/eth2/beacon_chain/req/data_column_sidecar_by_root/1/`

*[New in Deneb:EIP4844]*

The `<context-bytes>` field is calculated as `context = compute_fork_digest(fork_version, genesis_validators_root)`:

[1]: # (eth2spec: skip)

| `fork_version`           | Chunk SSZ type                |
|--------------------------|-------------------------------|
| `PEERDAS_FORK_VERSION`     | `peerdas.DataColumnSidecar`           |

Request Content:

```
(
  DataColumnIdentifier
)
```

Response Content:

```
(
  DataColumnSidecar
)
```

> The response is the column as `get_data_column_sidecar(signed_block: SignedBeaconBlock, blobs: Sequence[Blob], column_index: LineIndex)` computed.

Author: ```suggestion
```

  - [Helpers](#helpers)
      - [`verify_sample_proof_batch`](#verify_sample_proof_batch)
      - [`verify_data_column_sidecar_kzg_proof`](#verify_data_column_sidecar_kzg_proof)
      - [`verify_data_column_sidecar_inclusion_proof`](#verify_data_column_sidecar_inclusion_proof)
      - [`compute_subnet_for_data_column_sidecar`](#compute_subnet_for_data_column_sidecar)
  - [The gossip domain: gossipsub](#the-gossip-domain-gossipsub)
    - [Topics and messages](#topics-and-messages)
      - [Samples subnets](#samples-subnets)
        - [`data_column_sidecar_{subnet_id}`](#data_column_sidecar_subnet_id)
  - [The Req/Resp domain](#the-reqresp-domain)
    - [Messages](#messages)
      - [DataColumnSidecarByRoot v1](#datacolumnsidecarbyroot-v1)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->
<!-- /TOC -->

## Modifications in PeerDAS

### Preset

| Name                                     | Value                             | Description                                                         |
|------------------------------------------|-----------------------------------|---------------------------------------------------------------------|

> | `KZG_COMMITMENTS_INCLUSION_PROOF_DEPTH`   | `uint64(floorlog2(get_generalized_index(BeaconBlockBody, 'blob_kzg_commitments')))` (= 4) | <!-- predefined --> Merkle proof index for `blob_kzg_commitments` |

Reviewer 4: I noticed that there is also a preset named `KZG_COMMITMENT_INCLUSION_PROOF_DEPTH` in Deneb. The only difference is a letter 's'. I cannot tell the difference between the two. I think it's probably better to change the name of one or both of them so that we can tell the difference.

See https://github.com/ethereum/consensus-specs/blob/b2f210/specs/deneb/p2p-interface.md?plain=1#L56

### Configuration

| Name                                     | Value                             | Description                                                         |
|------------------------------------------|-----------------------------------|---------------------------------------------------------------------|

> | `DATA_COLUMN_SIDECAR_SUBNET_COUNT`              | `32`                               | The number of data column sidecar subnets used in the gossipsub protocol.  |

Reviewer 2: should the value be defined with respect to a previously defined constant like `NUMBER_OF_COLUMNS`?

Author: I'm not against it; just note that it is similar to the `BLOB_SIDECAR_SUBNET_COUNT`/`ATTESTATION_SUBNET_COUNT` definitions, and we didn't set them by `MAX_BLOBS_PER_BLOCK`/`MAX_ATTESTATIONS`.

I think at least we can add invariant `NUMBER_OF_COLUMNS % DATA_COLUMN_SIDECAR_SUBNET_COUNT == 0`.

### Containers

#### `DataColumnSidecar`

```python
class DataColumnSidecar(Container):
    index: LineIndex  # Index of column in extended matrix
    column: DataColumn
    kzg_commitments: List[KZGCommitment, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    kzg_proofs: List[KZGProof, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    signed_block_header: SignedBeaconBlockHeader
    kzg_commitments_inclusion_proof: Vector[Bytes32, KZG_COMMITMENTS_INCLUSION_PROOF_DEPTH]
```

#### `DataColumnIdentifier`

```python
class DataColumnIdentifier(Container):
    block_root: Root
    index: LineIndex
```

### Helpers

##### `verify_sample_proof_batch`

```python
def verify_sample_proof_batch(
        row_commitments: Sequence[KZGCommitment],
        row_ids: Sequence[LineIndex],
        column_ids: Sequence[LineIndex],
        datas: Sequence[Vector[BLSFieldElement, FIELD_ELEMENTS_PER_CELL]],
        proofs: Sequence[KZGProof]) -> bool:
    """
    Defined in polynomial-commitments-sampling.md
    """
    # pylint: disable=unused-argument
    ...
```

##### `verify_data_column_sidecar_kzg_proof`

```python
def verify_data_column_sidecar_kzg_proof(sidecar: DataColumnSidecar) -> bool:
    """

>     Verify if the commitment are correct

Reviewer: ```suggestion
    Verify if the proofs are correct
```
When we verify the commitments, we need the whole blobs. In this case, we don't have blobs and we have only the column, so we only verify the sample proofs.

    """
    row_ids = [LineIndex(i) for i in range(len(sidecar.column))]

>     column = sidecar.column
>     cell_count = len(column) // FIELD_ELEMENTS_PER_CELL
>     cells = [column[i * FIELD_ELEMENTS_PER_CELL:(i + 1) * FIELD_ELEMENTS_PER_CELL] for i in range(cell_count)]
>     assert len(cells) == len(sidecar.kzg_commitments) == len(sidecar.kzg_proofs)
> 
>     # KZG batch verify the cells match the corresponding commitments and proofs
>     return verify_sample_proof_batch(
>         row_commitments=sidecar.kzg_commitments,
>         row_ids=row_ids,  # all rows
>         column_ids=[sidecar.index],
>         datas=cells,
>         proofs=sidecar.kzg_proofs
>     )

Reviewer 5: ```suggestion
    assert len(sidecar.column) == len(sidecar.kzg_commitments) == len(sidecar.kzg_proofs)

    # KZG batch verifies that the cells match the corresponding commitments and proofs
    return verify_sample_proof_batch(
        row_commitments=sidecar.kzg_commitments,
        row_ids=row_ids,  # all rows
        column_ids=[sidecar.index],
        datas=sidecar.column,
        proofs=sidecar.kzg_proofs,
    )
```

```

##### `verify_data_column_sidecar_inclusion_proof`

```python
def verify_data_column_sidecar_inclusion_proof(sidecar: DataColumnSidecar) -> bool:
    """
    Verify if the given KZG commitments included in the given beacon block.
    """
    gindex = get_subtree_index(get_generalized_index(BeaconBlockBody, 'blob_kzg_commitments'))
    return is_valid_merkle_branch(
        leaf=hash_tree_root(sidecar.kzg_commitments),
        branch=sidecar.kzg_commitments_inclusion_proof,
        depth=KZG_COMMITMENTS_INCLUSION_PROOF_DEPTH,
        index=gindex,
        root=sidecar.signed_block_header.message.body_root,
    )
```

##### `compute_subnet_for_data_column_sidecar`

```python
def compute_subnet_for_data_column_sidecar(column_index: LineIndex) -> SubnetID:
    return SubnetID(column_index % DATA_COLUMN_SIDECAR_SUBNET_COUNT)
```

### The gossip domain: gossipsub

Some gossip meshes are upgraded in the fork of Pe to support upgraded types.

#### Topics and messages

##### Samples subnets

###### `data_column_sidecar_{subnet_id}`

This topic is used to propagate column sidecars, where each column maps to some `subnet_id`.

> The *type* of the payload of this topic is `DataColumn`.

Reviewer: ```suggestion
The *type* of the payload of this topic is `DataColumnSidecar`.
```

> The following validations MUST pass before forwarding the `sidecar: DataColumnSidecar` on the network, assuming the alias `block_header = blob_sidecar.signed_block_header.message`:

Reviewer: ```suggestion
The following validations MUST pass before forwarding the `sidecar: DataColumnSidecar` on the network, assuming the alias `block_header = sidecar.signed_block_header.message`:
```

- _[REJECT]_ The sidecar's index is consistent with `NUMBER_OF_COLUMNS` -- i.e. `sidecar.index < NUMBER_OF_COLUMNS`.

> - _[REJECT]_ The sidecar is for the correct subnet -- i.e. `compute_subnet_for_data_column_sidecar(blob_sidecar.index) == subnet_id`.

Reviewer 5: ```suggestion
- _[REJECT]_ The sidecar is for the correct subnet -- i.e. `compute_subnet_for_data_column_sidecar(sidecar.index) == subnet_id`.
```

- _[IGNORE]_ The sidecar is not from a future slot (with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` allowance) -- i.e. validate that `block_header.slot <= current_slot` (a client MAY queue future sidecars for processing at the appropriate slot).
- _[IGNORE]_ The sidecar is from a slot greater than the latest finalized slot -- i.e. validate that `block_header.slot > compute_start_slot_at_epoch(state.finalized_checkpoint.epoch)`

> - _[REJECT]_ The proposer signature of `blob_sidecar.signed_block_header`, is valid with respect to the `block_header.proposer_index` pubkey.

Reviewer 5: ```suggestion
- _[REJECT]_ The proposer signature of `sidecar.signed_block_header`, is valid with respect to the `block_header.proposer_index` pubkey.
```
